benches/galen.rs

```diff
@@ -0,0 +1,177 @@
+use dbsp::{
+    algebra::{FiniteMap, HasZero, ZSetHashMap},
+    circuit::{Root, Runtime, Stream},
+    operator::{CsvSource, DelayedFeedback},
+};
+
+use csv::{Reader as CsvReader, ReaderBuilder};
+
+use std::{fs::File, path::PathBuf};
+
+/*
+.decl p(X: Number, Z: Number)
+.decl q(X: Number, Y: Number, Z: Number)
+.decl r(R: Number, P: Number, E: Number)
+.decl c(Y: Number, Z: Number, W: Number)
+.decl u(R: Number, Z: Number, W: Number)
+.decl s(R: Number, P: Number)
+
+.input p(IO="file", filename="p.txt", delimiter=",")
+.input q(IO="file", filename="q.txt", delimiter=",")
+.input r(IO="file", filename="r.txt", delimiter=",")
+.input c(IO="file", filename="c.txt", delimiter=",")
+.input u(IO="file", filename="u.txt", delimiter=",")
+.input s(IO="file", filename="s.txt", delimiter=",")
+
+p(?x,?z) :- p(?x,?y), p(?y,?z).
+p(?x,?z) :- p(?y,?w), u(?w,?r,?z), q(?x,?r,?y).
+p(?x,?z) :- c(?y,?w,?z),p(?x,?w), p(?x,?y).
+q(?x,?r,?z) :- p(?x,?y), q(?y,?r,?z).
+q(?x,?q,?z) :- q(?x,?r,?z),s(?r,?q).
+q(?x,?e,?o) :- q(?x,?y,?z),r(?y,?u,?e),q(?z,?u,?o).
+*/
+
+type Number = u32;
+type Weight = isize;
+
+fn mk_csv_source<T>(fname: &str) -> CsvSource<File, ZSetHashMap<T, Weight>, T>
+where
+    T: Clone,
+{
+    let path: PathBuf = ["benches", "galen_data", fname].iter().collect();
+
+    let reader: CsvReader<File> = ReaderBuilder::new()
+        .delimiter(b',')
+        .has_headers(false)
+        .from_path(path)
+        .unwrap();
+    CsvSource::from_csv_reader(reader)
+}
+
+fn main() {
+    let hruntime = Runtime::run(1, |_runtime, _index| {
```

Comment:
I don't think this is a good idea.  Users should be able to change the number of workers without changing anything else in their code.  It is still possible to instantiate the circuit without a runtime though, in which case it will run on the caller's thread.

Reply:
This wouldn't really change anything else within their code, the first worker would always be spawned on the current thread so if the user always wanted the runtime to occur on another thread they'd just wrap it in a thread no matter what

Reply:
But what's the advantage?  Rust threads are light-weight enough that creating an extra one doesn't matter.

Reply:
Control, mostly. As an example, WASM doesn't support threads, so this would enable users on WASM to actually run dbsp (either in the browser or as part of a sandboxed plugin) which I think is something we should enable without forcing them to copy+paste the entire runtime

Reply:
WASM support is a valid argument. I'll create an issue for this.

Reply:
On a second thought, one shouldn't use `Runtime` if they only want one thread.  `Runtime` is only meant to provide services for multithreaded execution.  The circuit API is fully functional in a single-threaded context without a `Runtime`.